src/c/test.zig

```diff
@@ -1 +1,165 @@
+
+fn RequestContextType(comptime request_size: comptime_int) type {
+    return struct {
+        completion: *Completion,
+        sent_data: [request_size]u8 = undefined,
+        packet: *tb_packet_t = undefined,
+        result: ?struct {
+            context: usize,
+            client: tb_client_t,
+            packet: *tb_packet_t,
+            reply: ?[request_size]u8 = null,
+        } = null,
+
+        pub fn on_complete(
+            context: usize,
+            client: tb_client_t,
+            packet: *tb_packet_t,
+            result_ptr: ?[*]const u8,
+            result_len: u32,
+        ) callconv(.C) void {
+            var self = @intToPtr(*@This(), packet.user_data);
+            defer self.completion.complete();
+            self.result = .{
+                .context = context,
+                .client = client,
+                .packet = packet,
+            };
+
+            // Copy the message's body to the context buffer:
+            if (result_ptr != null and result_len > 0) {
+                assert(result_len == request_size);
+                self.result.?.reply = [_]u8{0} ** request_size;
+                std.mem.copy(u8, &self.result.?.reply.?, result_ptr.?[0..result_len]);
+            }
+        }
+    };
+}
+
+// Notifies the main thread when all pending requests are completed.
+const Completion = struct {
+    const Self = @This();
+
+    pending: usize,
+    mutex: Mutex = .{},
+    cond: Condition = .{},
+
+    pub fn complete(self: *Self) void {
+        self.mutex.lock();
+        defer self.mutex.unlock();
+
+        assert(self.pending > 0);
+        self.pending -= 1;
+        self.cond.signal();
+    }
+
+    pub fn wait_pending(self: *Self) void {
+        self.mutex.lock();
+        defer self.mutex.unlock();
+
+        while (self.pending > 0)
+            self.cond.wait(&self.mutex);
+    }
+};
+
+// When initialized with tb_client_echo_init, the c_client uses a test context that echoes the data back
+// without creating an actual client or connecting to a cluster.
+//
+// This same test should be implemented by all the target programming languages, asserting that:
+// 1. the c_client threading/signaling mechanism was initialized correctly.
+// 2. the client usage is correct, and they can submit and receive messages through the completion callback.
+// 3. the data marshaling is correct, and exactly the same data sent was received back, no matter the message length.
+test "c_client echo" {
+    const tb_completion_ctx: usize = 42;
+    var client: api.tb_client_t = undefined;
+    var packet_list: api.tb_packet_list_t = undefined;
+
+    // We ensure that the retry mechanism is being tested by allowing more simultaneous packets than "client_request_queue_max".
+    const num_packets = config.client_request_queue_max * 2;
+
+    // Using the create_accounts operation for this test.
+    const RequestContext = RequestContextType(@sizeOf(@import("../tigerbeetle.zig").Account));
```

Comment:
`ContextType` validates the request length based on the operation.
Not any particular reason to have chosen `create_accounts,` but in this case, the request must be a multiple of `@sizeOf(Account)`.

I just saw King's PR adding tests for `tb_client.h`, and with that we can change here to use the exported definitions from the c header.

Reply:
Could we change `RequestContextType(comptime request_size: comptime_int)` to `RequestContextType(comptime request_size_max: comptime_int)`? And instead of passing in the account size, pass in `config.message_body_size_max`.

- Then you can add a `sent_data_size` field (adjacent to `sent_data`) which would also determine the size of the reply.
- After generating the random data (`prng.random().bytes(&request.sent_data);`) you can also set a random `sent_data_size` (it just needs to be a multiple of the account size).

This will just add coverage of different message sizes.

Reply:
I rebased this branch and changed the test to use the c API directly from the `tb_client.h`.